language/fixtures/rescue.rb

```diff
@@ -1,4 +1,112 @@
 module RescueSpecs
+  class ClassVariableCaptor
+    def capture(msg)
+      @@captured_error = nil
+
+      raise msg
+    rescue => @@captured_error
+      :caught
+    end
+
+    def captured_error
+      @@captured_error
+    end
+  end
+
+  class ConstantCaptor
+    # Using lambda gets around the dynamic constant assignment warning
+    CAPTURE = -> msg {
+      begin
+        raise msg
+      rescue => CapturedError
+        :caught
+      end
+    }
+
+    def capture(msg)
+      self.class.remove_const(:CapturedError) if self.class.const_defined?(:CapturedError)
+
+      CAPTURE.call(msg)
+    end
+
+    def captured_error
+      CapturedError
+    end
+  end
+
+  class GlobalVariableCaptor
+    def capture(msg)
+      $captured_error = nil
+
+      raise msg
+    rescue => $captured_error
+      :caught
+    end
+
+    def captured_error
+      $captured_error
+    end
+  end
```

Comment:
Nitpicking - there is a global variable leaking. We try to minimize side effects of test cases.

Reply:
Yeah, I went back and forth on having a "teardown", but I wasn't sure it was worth it. Technically, the constant and class variable approaches also have side effects that last past the lifetime of their captor objects.

I'll put something together as a follow up commit.

Reply:
I see. Just FYI constants are tracked and there is a whitelist https://github.com/ruby/spec/blob/master/.mspec.constants

Reply:
Thanks for the link! Can you provide a little more context on the purpose of that list? It appears to only contain top level constants, but not all of them. For instance, I see `::SpecificExampleException` is included, but not `::RescueSpecs`.

Reply:
The list contains all the new the top-level constants introduced in the specs. The intention is to prevent declaring new classes/constants right into the test cases. It's implemented in `mspec` here https://github.com/ruby/mspec/blob/master/lib/mspec/runner/actions/constants_leak_checker.rb

To make life easier all the fixture classes (with suffix `Spec`) are allowed implicitly: https://github.com/ruby/spec/blob/master/default.mspec#L45-L48

Reply:
I see. Looks like no updates are needed to that file since all the constants introduced in this PR are nested under `RescueSpecs`, and none are top-level.

I did experiment with making it a top level constant though and used `CHECK_LEAKS=true` to ensure that the cleanup works as expected, which is does. 👍